### PR TITLE
WIP: Add Package Development guide

### PR DESCRIPTION
**Pull request by certik**
This PR is meant to summarize the discussion at https://github.com/pypa/python-packaging-user-guide/issues/320 in a way that everybody agrees on the summary written in this PR. Here is the plan:

1. In order to achieve an agreement, the document discusses 4 different ways to develop a Python package. It should be possible to agree on "best practices" for each way together with pros and cons.

2. Then we can discuss whether it is possible to agree on just one way that would be "officially" recommended.

It might not be possible to agree on 2., but it should be possible to agree on 1. This is a work in progress (WIP) PR, and I intend to update it with all the feedback that I can get. After we all agree on 1., we can discuss 2. and whether or not to merge this PR.

### CODE REVIEW

File: source/guides/package-development.rst [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+tests ``included``, ``src``
+===========================
+
+An example package in this cathegory is Matplotlib.
+
+tests ``separate``, ``non-src``
+===============================
+
+An example package in this cathegory is ?.
```
Comment (takluyver) [inline source/guides/package-development.rst]: Flit is an example. Many of the other packages I've made are also like this, e.g. [astcheck](https://github.com/takluyver/astcheck) is a very small example.

Comment (certik) [reply source/guides/package-development.rst]: Fixed in 9fcf77c081baa1a5932a3fcde8ecc442ac8a8726. Thanks!

**Comment by scolby33** [comment]
I disagree that this should be viewed as a matrix of `src`/`non-src` by `included`/`separate`. The original discussion was focused on `src`/`non-src` and drifted a bit. Additionally, I don't think I see a case where choosing a particular source layout changes the calculus related to the placement of your tests; that is to say, the decisions are truly independent modulo some insignificant path changes in your test or CI configuration. Is this intuition correct? If so, I would advocate for splitting the document into two--one for source layout and one for tests placement.

**Comment by certik** [comment]
@scolby33 thanks for the feedback. I reworked the layout in 53e6a3af22201a8b81674204d1e7a6b0a5326ba7, and it is now focused on the `src` / `non-src` issue, only within each variant there is the discussion of `included` vs `separate`.

Regarding your question whether the two are independent: they are not independent, the `separate` / `non-src` has the peculiar workaround of `rm -r flit` in order to test the installed package. The other 3 variants don't have that.

**Comment by takluyver** [comment]
Just to note: I wouldn't spend too much effort making this neatly presented, because I don't think it should become part of the packaging user guide. I see this as a tool for coordinating discussion, not something that will actually be merged.

**Comment by takluyver** [comment]
> the separate / non-src has the peculiar workaround of rm -r flit in order to test the installed package

I should clarify that I have never actually done this. When I'm developing a package, I use `flit install --symlink`, which is similar in effect to `pip install -e .`. So there's only one copy of the code to be tested.

**Comment by certik** [comment]
@takluyver yes, I it probably won't be merged.

@takluyver regarding `rm -r flit`, yes. The reason it is there is to show how to do it, since people were asking how to actually test the installed package (not in development mode).

**Comment by certik** [comment]
@takluyver I added instructions for flit in c07c3f8fa4df721fdfaf4df362ee14bf1a63bb30, and it seems just importing locally works, just like for sympy. So it seems for both flit and sympy one has the option to either import locally or install as a package?

Why do you install it as a package if you can just import locally?

**Comment by takluyver** [comment]
@certik I have it installed because I'm using it, and I want it to be importable without needing to be in the working directory. In the case of Flit, there's also a command line interface, which is only created when you install the package.

I have the development version installed, rather than the latest release, for dogfooding. I work on quite a few packages, and I don't usually switch between different environments for them. Instead I install all of them into `~/.local`, so that whatever I'm working on, I'm using the development version of other related packages, and I have a better chance of discovering issues before release. I use environments as well for specific purposes, but they're not part of my regular development workflow.

File: source/guides/package-development.rst [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+Pros:
+
+* Unlike the tests ``non-src`` case (both ``separate`` and ``included``), one
+  cannot accidentally run tests with the local package instead of the
+  installed one. environment
+
+Cons:
+
+* One can't import the package without installing it into an environment (what
```
[the remark this answers is not in the document]
Comment (certik) [reply source/guides/package-development.rst]: It's not a deal breaker for people who prefer this approach (since if that was a deal breaker, they would not be using this approach), but it's a cons for people who prefer a different setup. In this document I am trying to be completely neutral regarding a judgement: the "cons" of each of the 4 cases is judged from a perspective of somebody who prefers a different approach. The "pros" of each of the use cases is from somebody who prefers that very approach.

File: source/guides/package-development.rst [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+  version of ``flit``. One has to remove the ``flit`` directory with ``rm -r
+  flit`` to test the installed version.
+
+``src``
+=======
+
+tests ``included``
+------------------
+
+An example package in this cathegory is Matplotlib.
```
[the remark this answers is not in the document]
Comment (certik) [reply source/guides/package-development.rst]: Thanks, fixed in e41537f9f88aac31e9ba5c2b5f0502dac40abab5.

File: source/guides/package-development.rst [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+
+Repeat 1. and 2. The ``sympy`` environment only has the dependencies, it
+doesn't get modified and doesn't have the ``sympy`` package.
+
+Pros / Cons
+~~~~~~~~~~~
+
+Pros:
+
+* One can import the package locally, no need to install and pollute some
```
[the remark this answers is not in the document]
Comment (certik) [reply source/guides/package-development.rst]: If you want to import the package locally in the `src` setup --- can you provide the exact commands that do that?

I only know how to do that by adding the `src` directory into `PYTHONPATH`. I am assuming you have some scripts in the root (or `scripts` or `bin`) directory, that just import the package. Without adding it to `PYTHONPATH`, they will not work locally.

**Comment by certik** [comment]
@pradyunsg thank you for the feedback, much appreciated. I left comments above.

File: source/guides/package-development.rst [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+  environment
+
+Cons:
+
+* By installing the package and running ``pytest``, one will run the local
```
[the remark this answers is not in the document]
Comment (certik) [reply source/guides/package-development.rst]: Ok, I tried it and it doesn't work. Here is what I tried:
```
git clone https://github.com/takluyver/flit
conda create -y -n flit python=3.7 pytest requests requests_download testpath responses docutils pytoml pytest-cov
cd flit
conda activate flit
mkdir tmp
cd tmp
pytest ../tests/
```
and I got the following output: https://gist.github.com/certik/afb931f4f72ad171cab83b9141cdcfcc, there are some failures, but for example the test `tests/test_build.py` executes without problems, and that file contains the line:
```
from flit import build, common
```
So flit got imported, even though if I import `flit` in the `tmp` directory, it will not be found:
```
$ python -c "import flit"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'flit'
```

Conclusion: you can't "cd and run tests from a different directory". This will happily import the local `flit` anyway, which specifically you do not want.

@takluyver do you agree with this analysis, or have I done some mistake?

**Comment by certik** [comment]
@pradyunsg I would be interested what you think of my comments above.

File: source/guides/package-development.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (takluyver) [reply source/guides/package-development.rst]: I don't use this technique myself, so I'm not very familiar with it, but I see there is a `--rootdir` option for pytest; maybe that affects where it will find imports.

File: source/guides/package-development.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+doesn't get modified and doesn't have the ``sympy`` package.
+
+Pros / Cons
+~~~~~~~~~~~
+
+Pros:
+
+* One can import the package locally, no need to install and pollute some
+  environment
+* One can still install the package to test it if one wants to
```
[the remark this answers is not in the document]
Comment (takluyver) [reply source/guides/package-development.rst]: I think part of the reason I don't worry about testing with the installed package is that I don't tend to run into this. There are two main ways that what you install can be different from your source:

1. If there is a deliberate build step of some kind (e.g. compiling C code, minifying Javascript...), which my own projects tend not to do.
2. If you miss files out of the manifest or package_data lists, but I consider this a design flaw in distutils/setuptools, not an inevitable part of packaging. I'm not claiming to have it totally solved, but I think it's harder to accidentally mess this up with Flit.

**Comment by pradyunsg** [comment]
I'm currently working with a fair amount of life stuff happening, enough to not have time for putting time into OSS.

I'll come around to this discussion once things settle down a bit for me. :)

File: source/guides/package-development.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (uranusjr) [reply source/guides/package-development.rst]: Maybe it’s possible to reword this as “one can import the package directly without extra tooling or configuration”? The current wording implies (to me) that a user *cannot* import a package in `src` without installing it, which is not true (because e.g. `PYTHONPATH`).

[the remark this answers is not in the document]
Comment (certik) [reply source/guides/package-development.rst]: @uranusjr good point. I reworded it based on your suggestion in 69dac7a81617a16936533c822351439f103a36fc. Let me know what you think.

**Comment by theacodes** [comment]
I'm unsure what to do with this PR right now. @certik, you've provided us with incredible content that has driven the discussion around #320, but I don't think we have enough consensus to accept a *guide*. I know that @pradyunsg is working on a *discussion* that will probably be a better venue for this content.

So I'm curious if we should close this PR and open a new one for you a @pradyunsg to work on putting this into a guide, so we're not encumbered by the comments and discussion here.

**Comment by pradyunsg** [comment]
I'll file my PR in a few days -- I got caught up in academics and almost all of my OSS time went into prepping for the next pip release.

**Comment by theacodes** [comment]
No worries, thanks for the update!

**Comment by certik** [comment]
@theacodes I haven't found time to address @pfmoore's review. Some of the points would require us to have a video conversation about the issues, so that we can arrive at some kind of an agreement about the pros and cons of each approach --- I am afraid github comments are not a good medium for that. Unfortunately I am really busy right now to pursue that approach.

So given that, and also the fact that my opinion is in minority of the people actively participating in here, I am going to go ahead and close the PR. But I would like to encourage all of you to please actively seek the opinion of the people who do not participate here, and ensure that the majority likes your eventual decision about this.
